### ISSUE TITLE
[Gnome 46]: switching to a workspace makes the non-contrl key "stick"

### ISSUE THREAD

**Issue by vladzcloudius** [mzur/gnome-shell-wsmatrix#279]
## Configuration
OS: Ubuntu 24.04
Grid config: 2x4
Extension version: 44

## Description
I use `Ctrl+Alt+Arrow X` keybinding to switch between workspaces.
With U24.04 when I switch to a workspace in the direction X the "Arrow X" key "sticks": e.g. if a destination workspace has a text editor open then the cursor is going to move in the corresponding direction.

I verified that this has nothing to do with arrow keys: I configured a `Ctrl+Alt+m` to be a shortcut to switch to the workspace above and the issue reproduced: letter 'm' began to be printed endlessly until I press another key on the keyboard.

**Comment by vladzcloudius** [comment]
cc @mzur

**Comment by andersjohansson** [comment]
I also experience this, but I have only seen it in Signal (an electron application) so I thought it was a bug there. Yes, tested in chromium as well, and it appears there as well (about every third workspace switch), so perhaps it is some kind of interaction between chrome/electron and wayland/gnome that goes wrong?

**Comment by matttbe** [comment]
Same here: I have the same issue with Thunderbird and Firefox, but not with "native" Gnome apps like the terminal, nautilus and the text editor.

(Great extension BTW, hard to keep it disabled :) )

**Comment by vmarchaud** [comment]
I've tested around few options and it seems that when disabling the popup it doesn't trigger the bug if that can helps narrow down the issue

**Comment by mzur** [comment]
I'll probably have a look once I migrated to GNOME Shell 46 myself.

**Comment by vladzcloudius** [comment]
> I've tested around few options and it seems that when disabling the popup it doesn't trigger the bug if that can helps narrow down the issue

Yup, that's how I worked this around as well: I set "time to show the popup" to `1` and it stopped reproducing.

**Comment by longwave** [comment]
Also seeing this since upgrading to Ubuntu 24.04 with GNOME 46, disabling the popup works for me too.

**Comment by alexbrouwer** [comment]
> > I've tested around few options and it seems that when disabling the popup it doesn't trigger the bug if that can helps narrow down the issue
> 
> Yup, that's how I worked this around as well: I set "time to show the popup" to `1` and it stopped reproducing.

Me too. Upgraded to Ubuntu 24.04 with GNOME 46. Same issue. In my case mostly when navigating to the workspace with my IDE (Intellij) on it. Disabling the popup seems to work as a workaround.

Love to see a solution to make my favorite extension 100% usable again!

**Comment by tmortagne** [comment]
I finally notice this mess after upgrading to Ubuntu 24.04 was happening shortly after switching from one workspace to another and found this issue. Thanks for the explanation @vladzcloudius.

This problem unfortunately makes this extension unusable, so I hope it will get fixed. I'm a bit lost without my grid of workspaces.

**Comment by gurupras** [comment]
I am running into this _far too much_ and am willing to put in some effort to track this down.
But I have never worked on GNOME extensions before. If you can give me a brief set of instructions on how I can go about investigating this, I am happy to do run through the steps and report back.

**Comment by vmarchaud** [comment]
> I am running into this _far too much_ and am willing to put in some effort to track this down. But I have never worked on GNOME extensions before. If you can give me a brief set of instructions on how I can go about investigating this, I am happy to do run through the steps and report back.
> 
> FWIW, I tried the "time to show the popup" period of 1s and that didn't make the problem go away. I was originally on 500ms.

There are some instructions over the readme https://github.com/mzur/gnome-shell-wsmatrix#developing

**Comment by vmarchaud** [comment]
I've tried to investigate a bit although i never worked on gnome extensions before:
- The problem lies not in the popup itself but the fact that we are switching more than one workspace at once which trigger the bug, just going to right (or left) and back is enough.
- By "once" i mean triggering a switch *while* already switching and the popup is still visible, a mean to verify this is to reduce the time the popup appears to < 100ms and the bug will be way harder to reproduce.
- From my debugging, it seems taht the first "switch" trigger this [code path](https://github.com/mzur/gnome-shell-wsmatrix/blob/master/wsmatrix%40martin.zurowietz.de/workspacePopup/workspaceManagerOverride.js#L428) but i'm can't find from where (there is only one place that this function is called and i also had a log that didn't print), while the second trigger [this one](https://github.com/mzur/gnome-shell-wsmatrix/blob/master/wsmatrix%40martin.zurowietz.de/workspacePopup/workspaceSwitcherPopup.js#L173) which calls the previous mentioned function and then trigger the bug.

I've also looked at @mzur changes for gnome 46 and i don't see (obviously from someone that never saw a gnome extension previously so to take with a pinch of salt) any changes that would explain this so i suspect this comes from gnome itself. Can't go any further as i have no idea how to continue debugging :/

**Comment by vakker** [comment]
Same issue, I'm glad it's not just me!
One thing to add, I found `xev` very helpful to debug, it shows specifically which keys are registered by the system.
For the time being, disabling the pop-up works around this.

**Comment by Atko** [comment]
Same here: Ubuntu 24.04, Gnome 46, disabling popup solves the "sticky key" issue.

**Comment by vmarchaud** [comment]
One of my coworker just had a similar issue where the ctrl key was blocked, however he wasn't using wsmatrix so it may well be a gnome issue

**Comment by itsrachelfish** [comment]
Using Ubuntu 24.04 on a brand new Framework laptop and I was able to replicate this issue.

Thanks for the suggestion for disabling the popup in workspace matrix. I like seeing the preview, but keys getting stuck is a major hassle.

**Comment by stelian42** [comment]
Same here, same comments as @itsrachelfish

**Comment by ebeem** [comment]
There was a race condition that I believe still exists in grabbing the release of the modifier key (like `alt`).
My previous commits didn't properly comment such behaviors, I think the code was simplified later but this race condition was not taken into account.

Review more about the issue here: https://bugzilla.gnome.org/show_bug.cgi?id=596695
If someone is willing to test and fix it (I am not 100% sure this is actually the problem, it's just a guess), you can try reverting the commit here dc813d7.

Unfortunately, I am not using gnome anymore so I can't test myself, but thought I might be able to give a hint.

**Comment by scottkosty** [comment]
I experience this issue, as described by others.

Can someone give me an ELI5 version of how to test  @ebeem's theory (just above this comment)? I know nothing about gnome extensions. I see the following file on my computer: `/home/scott/.local/share/gnome-shell/extensions/wsmatrix@martin.zurowietz.de/workspacePopup/workspaceSwitcherPopup.js`. Is it sufficient to edit that file? How do I then reinitialize the extension without restarting my computer so I can quickly test?

**Comment by scottkosty** [comment]
Disabling the popup indeed works, but if anyone prefers, I found that setting the "time to show the the popup" to 50 ms also eliminated the issue and still gives me at least an idea of where I am in my messing workspaces.

**Comment by mzur** [comment]
@scottkosty Thanks for testing! You can just edit the file and restart GNOME (explained [here](https://github.com/mzur/gnome-shell-wsmatrix?tab=readme-ov-file#developing)).

**Comment by scottkosty** [comment]
Thanks! I tested the proposed change but can still reproduce the issue with the change.

**Comment by sandroden** [comment]
Same problems as described above. To me the [proposed solution](https://github.com/mzur/gnome-shell-wsmatrix/issues/279#issuecomment-2228214450)  solved the issue. Thanks a lot. 
I'm using Ubuntu 24.04 with wayland.
I also noticed that Control key was not release only if I double hit the arrows fast enough. Doing it slowly didn't show the issue.

**Comment by vmarchaud** [comment]
I've been running Ubuntu 24.10 for about 3 weeks now with Gnome 47 and wsmatrix v11.1 and the issue seems to have been fixed, can anyone else confirm ?

**Comment by tmortagne** [comment]
> can anyone else confirm ?

I just put back the popup after seeing your message, and I indeed cannot reproduce this problem on Ubuntu 24.10.

**Comment by vladzcloudius** [comment]
I'm on 24.04 and it doesn't reproduce (easily) either. I'll update here if it does.
Was there a fix that explains this? If "yes" then let's reference it and close.
We can always re-open if it reproduces.

**Comment by tmortagne** [comment]
Maybe something was modified in gnome-shell or mutter.

**Comment by vladzcloudius** [comment]
Ok. Closing...

Packages version I have:

```
sudo dpkg -l | egrep "gnome-shell|mutter"                                                                                                                           
ii  chrome-gnome-shell                            42.1-4                                   all          GNOME Shell integration for browsers - transitional package
ii  gir1.2-mutter-14:amd64                        46.2-1ubuntu0.24.04.4                    amd64        GObject introspection data for Mutter
ii  gnome-shell                                   46.0-0ubuntu6~24.04.5                    amd64        graphical shell for the GNOME desktop
ii  gnome-shell-common                            46.0-0ubuntu6~24.04.5                    all          common files for the GNOME graphical shell
ii  gnome-shell-extension-appindicator            58-1ubuntu24.04.1                        all          AppIndicator, KStatusNotifierItem and tray support for GNOME Shell
ii  gnome-shell-extension-desktop-icons-ng        46+really47.0.9-1                        all          desktop icon support for GNOME Shell
ii  gnome-shell-extension-prefs                   46.0-0ubuntu6~24.04.5                    amd64        tool to enable / disable GNOME Shell extensions
ii  gnome-shell-extension-ubuntu-dock             90ubuntu2                                all          Ubuntu Dock for GNOME Shell
ii  gnome-shell-extension-ubuntu-tiling-assistant 46-1ubuntu1.1                            all          extension which adds a Windows-like snap assist to GNOME Shell
ii  gnome-shell-extensions                        46.1-2                                   all          Extensions to extend functionality of GNOME Shell
ii  libmutter-14-0:amd64                          46.2-1ubuntu0.24.04.4                    amd64        window manager library from the Mutter window manager
ii  mutter-common                                 46.2-1ubuntu0.24.04.4                    all          shared files for the Mutter window manager
ii  mutter-common-bin                             46.2-1ubuntu0.24.04.4                    amd64        shared programs for the Mutter window manager
ii  yaru-theme-gnome-shell                        24.04.2-0ubuntu1                         all          Yaru GNOME Shell desktop theme from the Ubuntu Community
```

Full list:
[packages.txt.gz](https://github.com/user-attachments/files/17937937/packages.txt.gz)

**Comment by tmortagne** [comment]
Bad news, I just reproduced it. It's definitely a lot rarer (took me a whole day of constantly switching workspaces), but it still happens, apparently.

**Comment by vladzcloudius** [comment]
> Bad news, I just reproduced it. It's definitely a lot rarer (took me a 2 days of constantly switching workspaces), but it still happens, apparently (still, it's good enough for me 🙂).

On my end it doesn't reproduce yet.
Reopening.

**Comment by itsrachelfish** [comment]
@tmortagne @vladzcloudius I am able to reproduce this bug reliably on Ubuntu v24.04 with Gnome v46.0. I believe this bug is chromium specific, I've been able to reproduce it using chromium and chromium derivatives such as electron apps. I tested with packages installed via snap as well as apt, the behavior is the same in both cases.

Examples of impacted programs that I've tested:
- chromium (snap package)
- vscode (snap package)
- keybase (apt package)

Both vscode and keybase are chromium based electron apps. Signal desktop and Element are other electron apps you could try, but I didn't test those.

Programs I've tested which are NOT impacted:
- firefox
- geddit
- gnome-terminal
- nautilus

All you need to do to reproduce this bug is use the keyboard hotkeys (ctrl + alt + arrow keys) and switch into chromium with a text editor highlighted. Your cursor will move around like crazy. If you press "alt" your browser will go backwards or forwards through your page history (as if you were pressing "alt + arrow key"). In vscode you can click and multiple cursors will be created (as if you were pressing "alt + click")

**Comment by mzur** [comment]
Thanks for the info @itsrachelfish! I'm (finally) using GNOME 46.0 on Ubuntu 24.04, too, but unfortunately I could not reproduce this behavior yet, be it with Chromium or something else. What are your exact extension settings?

**Comment by itsrachelfish** [comment]
@mzur I just reformatted my computer yesterday, installed Ubuntu, and installed the workspace matrix extension. The bug happened for me with default settings. I worked around the issue by disabling the popup.

Interestingly after updating all of my apt packages and rebooting my system it seems like I am no longer able to reproduce this bug even with the workspace popup enabled.

Perhaps it was an upstream wayland bug after all? :thinking:
